ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryUrlSigningService.java

```diff
@@ -0,0 +1,138 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.frontend;
+
+import com.github.ambry.rest.RestMethod;
+import com.github.ambry.rest.RestRequest;
+import com.github.ambry.rest.RestServiceErrorCode;
+import com.github.ambry.rest.RestServiceException;
+import com.github.ambry.rest.RestUtils;
+import com.github.ambry.utils.Time;
+import java.util.Map;
+
+
+/**
+ * Default implementation of {@link UrlSigningService} that currently only converts all POST headers into query
+ * parameters and does not actually reliably sign the URL.
+ * <p/>
+ * A real implementation needs to encode enough information for manipulations to be discovered.
+ */
+public class AmbryUrlSigningService implements UrlSigningService {
+  static final String AMBRY_PARAMETERS_PREFIX = "x-ambry-";
+  private static final String LINK_EXPIRY_TIME = "et";
+  private static final String ENDPOINT_SUFFIX = "/";
+  private static final String QUERY_STRING_START = "?";
+  private static final String PARAMETER_SEPARATOR = "&";
+  private static final String PARAMETER_ASSIGN = "=";
+
+  private final String uploadEndpoint;
+  private final String downloadEndpoint;
+  private final long defaultUrlTtlSecs;
+  private final long defaultMaxUploadSize;
+  private final long maxUrlTtlSecs;
+  private final Time time;
+
+  AmbryUrlSigningService(String uploadEndpoint, String downloadEndpoint, long defaultUrlTtlSecs,
+      long defaultMaxUploadSize, long maxUrlTtlSecs, Time time) {
+    if (!uploadEndpoint.endsWith(ENDPOINT_SUFFIX)) {
+      uploadEndpoint = uploadEndpoint + ENDPOINT_SUFFIX;
+    }
+    if (!downloadEndpoint.endsWith(ENDPOINT_SUFFIX)) {
+      downloadEndpoint = downloadEndpoint + ENDPOINT_SUFFIX;
+    }
+    this.uploadEndpoint = uploadEndpoint;
+    this.downloadEndpoint = downloadEndpoint;
+    this.defaultUrlTtlSecs = defaultUrlTtlSecs;
+    this.defaultMaxUploadSize = defaultMaxUploadSize;
+    this.maxUrlTtlSecs = maxUrlTtlSecs;
+    this.time = time;
+  }
+
+  @Override
+  public String getSignedUrl(RestRequest restRequest) throws RestServiceException {
+    Map<String, Object> args = restRequest.getArgs();
+    String restMethodInSignedUrlStr = RestUtils.getHeader(args, RestUtils.Headers.URL_TYPE, true);
+    RestMethod restMethodInSignedUrl;
+    try {
+      restMethodInSignedUrl = RestMethod.valueOf(restMethodInSignedUrlStr);
+    } catch (IllegalArgumentException e) {
+      throw new RestServiceException("Unrecognized RestMethod: " + restMethodInSignedUrlStr,
+          RestServiceErrorCode.InvalidArgs);
+    }
+    StringBuilder urlBuilder = new StringBuilder();
```

Comment:
I wonder if there is UriBuilder in the standard library or some library that we already depend on?

Reply:
I looked around a bit (not hard enough TBH) but figured that doing it myself was easy and short - there isn't any complicated code and since we pick stuff from a URL we have received, everything must be URL safe.